Documentation/Evolution/StronglyTypedCaptures.md

```diff
@@ -0,0 +1,684 @@
+//         Optionally {
+//             ".."
+//             OneOrMore(CharacterClass.hexDigit).capture()
+//         }
+//     }
+//     // `.Captures == (Substring, Substring?)`
+```
+
+#### Nested capturing group
+
+When capturing group is nested within another capturing group, they count as two
+distinct captures in the order their left parenthesis first appears in the
+regular expression literal. This is consistent with PCRE and allows us to use
+backreferences (e.g. `\2`) with linear indices.
+
+```swift
+let graphemeBreakPropertyData = /(([0-9A-F]+)(\.\.([0-9A-F]+)))\s*;\s(\w+).*/
+// Positions in capture tuple: 0 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~    4 ^~~~~
+//                                         2 ^~~~~~~~~~~~~~~~~
+//                              1 ^~~~~~~~~~~   3 ^~~~~~~~~~~
+// => `Regex<(Substring, Substring, Substring, Substring, Substring)>`
+
+// Result builder DSL equivalent:
+//     let graphemeBreakPropertyData = Regex {
+//         Regex {
+//             OneOrMore(CharacterClass.hexDigit).capture() // (1)
+//             Regex {
+//                 ".."
+//                 OneOrMore(CharacterClass.hexDigit).capture() // (3)
+//             }.capture() // (2)
+//         }.capture() // (0)
+//         Repeat(CharacterClass.whitespace)
+//         ";"
+//         CharacterClass.whitespace
+//         OneOrMore(CharacterClass.word).capture() // (4)
+//         Repeat(CharacterClass.any)
+//     }
+//     // `.Captures == Regex(Substring, Substring, Substring, Substring, Substring)`
+
+let input = "007F..009F    ; Control # Cc  [33] <control-007F>..<control-009F>"
+// Match result for `input`:
+// ("007F..009F", "007F", "..009F", "009F", "Control")
+```
+
+
+#### Quantification
+
+A quantifier's capture type depends on its underlying pattern. The laziness of
+a quantifier, as denoted by a `?` suffix, does not affect the capture type. 
+
+```
+pattern ::= pattern '*' '?'?
+pattern ::= pattern '+' '?'?
+pattern ::= pattern '?' '?'?
+pattern ::= pattern '{' number (',' spaces? number?)? '}' '?'?
+```
+
+| Syntax               | Description           | Capture type                                                  |
+| -------------------- | --------------------- | ------------------------------------------------------------- |
+| `*` / `*?`           | 0 or more             | `Array` of the sub-pattern capture type                       |
+| `+` / `+?`           | 1 or more             | `Array` of the sub-pattern capture type                       |
+| `?` / `??`           | 0 or 1                | `Optional` of the sub-pattern capture type                    |
+| `{n}` / `{n}?`       | Exactly _n_           | `Array` of the sub-pattern capture type                       |
+| `{n, m}` / `{n, m}?` | Between _n_ and _m_   | `Array` of the sub-pattern capture type                       |
+| `{n,}` / `{n,}?`     | _n_ or more           | `Array` of the sub-pattern capture type                       |
```

Comment:
Quantification is more complex if we're talking about syntax or capabilities (e.g. possessive quantification, whether `{n}?` means anything anymore, etc.). The key point is that all of this is entirely irrelevant to this pitch.

---

#### Quantification: `*`, `+`, `?`, `{n}`, `{n,}`, `{n,m}`

A quantifier wraps its underlying pattern's capture type in either an `Optional` or `Array`. Zero-or-one quantification (`?`) produces an `Optional` and all others produce an `Array`. The kind of quantification, i.e. greedy vs reluctant vs possessive, is irrelevant to determining the capture type.

...

Reply:
FWIW, the token representation from my [draft](https://github.com/apple/swift-experimental-string-processing/pull/40/files#diff-21e1fabf61ddc50761ff35e58d9df816058df5224058a64ffb93ad4f7ae7c58cR162):

```swift
  enum Quantifier {
    case zeroOrMore(Kind)   // *,     *?,     *+
    case oneOrMore(Kind)    // +,     +?,     ++
    case zeroOrOne(Kind)    // ?,     ??,     ?+
    case exactly(Int, Kind) // {n},   {n}?,   {n}+
    case nOrMore(Int, Kind) // {n,},  {n,}?,  {n,}+
    case upToN(Int, Kind)   // {,n},  {,n}?,  {,n}+
    case range(             // {n,m}, {n,m}?, {n,m}+
      atLeast: Int, atMost: Int, Kind)

    enum Kind {
      case greedy     //
      case reluctant  // ?
      case possessive // +
    }
  }
```

PCRE doesn't have `{,n}`, and reluctant or possessive quantification with `{n}` and variants gets weird pretty quickly. [Oniguruma](https://github.com/kkos/oniguruma/blob/master/doc/RE#L154) doesn't typically support `{n}?` as meaning reluctant, as that's kinda meaningless, unless you specify/configure with `ONIG_SYNTAX_JAVA` or `ONIG_SYNTAX_PERL`. Otherwise, the `?` changes meaning from reluctant to zero-or-one. Similarly for some of the possessives.

We really don't want to wade into this in this pitch :-). CC @hamishknight

Reply:
Updated as suggested.